types/react-relay/test/relay-experimental-tests.tsx

```diff
@@ -880,3 +881,49 @@ function Subscription() {
         return <p>The subscription has been started in the background.</p>;
     };
 }
+
+/**
+ * Tests for useQueryLoader
+ * see https://relay.dev/docs/en/experimental/api-reference#usequeryloader
+ */
+function QueryLoader() {
+    const query = graphql`
+        query AppQuery($id: ID!) {
+            user(id: $id) {
+                name
+            }
+        }
+    `;
+
+    const [queryReference, loadQuery, disposeQuery] = useQueryLoader<AppQuery>(query);
+    type queryReferenceType = typeof queryReference;
```

Comment:
This seems weird, can you update this test to add a copy of the typing that relay-compiler and the TS plugin emit for this and use that instead?

Reply:
In our code the way we would normally type `queryReference` is:

```ts
import { PreloadedQuery } from "react-relay/lib/relay-experimental/EntryPointTypes"
import { Graphql_SomeQuery } from "services/__generated__/Graphql_SomeQuery.graphql"

const Component: React.FC<{ queryReference: PreloadedQuery<Graphql_SomeQuery> }> = ({ queryReference }) => {
  // ...
}
```

Would it work if I exported `PreloadedQuery` from `react-relay/hook.d.ts`? In that case the above test example would be simplified down to:

```ts
import { PreloadedQuery } from "react-relay/hooks"

type queryReferenceType = PreloadedQuery<AppQuery>

// ...

function NameDisplay({ queryReference }: { queryReference: queryReferenceType }) {

//...

```

Reply:
I’m not sure I follow why it matters where the type is imported from (it should be exported only where the upstream package does it). The thing I’m mostly missing in this test is the`AppQuery` typings that would normally exist in that `__generated__` artefact.

Reply:
The `AppQuery` type (and subtypes) were previously defined in the test file [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2a0e1a08928a45d631767696662e2deaa92012ad/types/react-relay/test/relay-experimental-tests.tsx#L72). These look like type generated based on the examples in [the docs](https://relay.dev/docs/en/experimental/api-reference#uselazyloadquery). Said types were then used in the existing `PreloadedQuery` and `LazyLoadQuery` test, so I just re-used it for the `QueryLoader` test since all three of these tests reproduces the functionality described in the docs.

Is this sufficient for the PR, or should I generate a different set of types for this test case?

In terms of exports, the upstream package exports all types from `EntryPointTypes.flow` right [here](https://github.com/facebook/relay/blob/master/packages/relay-experimental/index.js#L40), which is then exported through the `react-relay/hooks` entry point. It looks like that behavior never got carried over to the TS types, so if the goal is alignment with the upstream flow types then I will update this PR with the missing exports.

Reply:
I updated the types to reflect the upstream exports, and change up the test to be more representative of a real-world use case. I'm still not sure whether to change anything about the `AppQuery` type since it already looks like a generated relay type.